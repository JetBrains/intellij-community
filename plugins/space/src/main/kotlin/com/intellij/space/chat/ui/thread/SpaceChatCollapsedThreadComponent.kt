// Copyright 2000-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.space.chat.ui.thread

import circlet.m2.threads.M2ThreadPreviewVm
import circlet.platform.api.toKDateTime
import circlet.platform.client.resolve
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.text.HtmlChunk
import com.intellij.space.chat.model.api.SpaceChatItem
import com.intellij.space.chat.ui.SpaceChatAvatarType
import com.intellij.space.messages.SpaceBundle
import com.intellij.space.stats.SpaceStatsCounterCollector
import com.intellij.space.ui.SpaceAutoUpdatableComponentService
import com.intellij.space.ui.SpaceAvatarProvider
import com.intellij.space.utils.formatPrettyDateTime
import com.intellij.ui.components.JBLabel
import com.intellij.ui.components.labels.LinkLabel
import com.intellij.ui.components.panels.HorizontalLayout
import com.intellij.util.ui.JBUI
import com.intellij.util.ui.UIUtil
import com.intellij.util.ui.components.BorderLayoutPanel
import libraries.coroutines.extra.Lifetime
import libraries.coroutines.extra.launch
import runtime.Ui
import runtime.reactive.Property
import runtime.reactive.SequentialLifetimes
import javax.swing.JComponent
import javax.swing.JPanel

internal fun createCollapsedThreadComponent(
  project: Project,
  lifetime: Lifetime,
  message: SpaceChatItem,
  threadActionsFactory: SpaceChatThreadActionsFactory?,
  withFirst: Boolean = true
): JComponent {
  val contentPanel = BorderLayoutPanel().apply {
    isOpaque = false
    border = JBUI.Borders.emptyTop(8)
  }
  val threadPreview = message.threadPreview
  val loadingThreadsSqLifetime = SequentialLifetimes(lifetime)
  message.loadingThread.forEach(lifetime) { loadingThread ->
    val loadingThreadLifetime = loadingThreadsSqLifetime.next()
    contentPanel.removeAll()
    if (loadingThread != null) {
      if (!loadingThread.isCompleted) {
        val loadingThreadLabel = JBLabel(SpaceBundle.message("chat.message.loading.thread.label")).apply {
          foreground = UIUtil.getContextHelpForeground()
        }
        contentPanel.addToCenter(loadingThreadLabel)
        contentPanel.revalidate()
        contentPanel.repaint()
      }
      launch(lifetime, Ui) {
        val threadValue = loadingThread.await()
        val threadComponent = createThreadComponent(
          project, lifetime, threadValue,
          pendingStateProvider = { false },
          threadActionsFactory = threadActionsFactory,
          folded = false,
          withFirst = withFirst
        )
        threadValue.mvms.forEach(loadingThreadLifetime) { messages ->
          contentPanel.isVisible = messages.messages.size > if (withFirst) 0 else 1
        }
        contentPanel.removeAll()
        contentPanel.addToCenter(threadComponent)
        contentPanel.revalidate()
        contentPanel.repaint()
      }
    }
    else {
      if (threadPreview != null) {
        val actionsComponent = createActionsComponent(project, lifetime, message, threadPreview, threadActionsFactory)
        contentPanel.addToCenter(actionsComponent)
        threadPreview.messageCount.forEach(loadingThreadLifetime) { count ->
          contentPanel.isVisible = count > 0
        }
      }
      else {
        contentPanel.isVisible = false
      }
      contentPanel.revalidate()
      contentPanel.repaint()
    }
  }
  return contentPanel
}

private fun createActionsComponent(
  project: Project,
  lifetime: Lifetime,
  message: SpaceChatItem,
  threadPreview: M2ThreadPreviewVm,
  threadActionsFactory: SpaceChatThreadActionsFactory?
): JComponent {
  val authorsPanel = createThreadAuthorAvatarsComponent(lifetime, threadPreview)
  val dateComponent = createDateComponent(lifetime, threadPreview.lastReplyTime)
  val repliesComponent = createRepliesLink(project, lifetime, threadPreview.messageCount) {
    launch(lifetime, Ui) {
      message.loadThread(lifetime)
    }
  }
  return JPanel(HorizontalLayout(JBUI.scale(5))).apply {
    isOpaque = false
    add(authorsPanel)
    add(repliesComponent)
    add(dateComponent)
    threadActionsFactory?.createActionsComponent()?.let { add(it) }
  }
}

private fun createThreadAuthorAvatarsComponent(lifetime: Lifetime, threadPreview: M2ThreadPreviewVm): JComponent {
  val contentPanel = BorderLayoutPanel().apply {
    isOpaque = false
  }
  val avatarsProvider = SpaceAvatarProvider(lifetime, contentPanel, SpaceChatAvatarType.THREAD.size)
  threadPreview.authors.forEach(lifetime) { authors ->
    contentPanel.removeAll()
    if (authors.isNullOrEmpty()) {
      contentPanel.border = JBUI.Borders.emptyLeft(10)
    }
    else {
      contentPanel.border = JBUI.Borders.empty()
      val icons = authors.take(3).map { avatarsProvider.getIcon(it.resolve()) }
      val iconsPanel = JPanel(HorizontalLayout(1)).apply {
        isOpaque = false
        for (icon in icons) {
          add(JBLabel(icon))
        }
      }
      contentPanel.addToCenter(iconsPanel)
      contentPanel.revalidate()
      contentPanel.repaint()
    }
  }
  return contentPanel
}

private fun createDateComponent(lifetime: Lifetime, lastReplyTime: Property<Long?>): JComponent {
  val contentPanel = BorderLayoutPanel().apply {
    isOpaque = false
  }
  lastReplyTime.forEach(lifetime) { time ->
    contentPanel.removeAll()
    if (time == null) {
      contentPanel.isVisible = false
    }
    else {
      contentPanel.isVisible = true
      val dateComponent = SpaceAutoUpdatableComponentService.createAutoUpdatableComponent {
        JBLabel(HtmlChunk.text(time.toKDateTime().formatPrettyDateTime()).toString()).apply { // NON-NLS
          putClientProperty(UIUtil.HIDE_EDITOR_FROM_DATA_CONTEXT_PROPERTY, true)
          foreground = UIUtil.getContextHelpForeground()
          setCopyable(true)
        }
      }
      contentPanel.addToCenter(dateComponent)
    }
    contentPanel.revalidate()
    contentPanel.repaint()
  }
  return contentPanel
}

private fun createRepliesLink(project: Project, lifetime: Lifetime, messagesCount: Property<Int>, action: () -> Unit): JComponent {
  val contentPanel = BorderLayoutPanel().apply {
    isOpaque = false
  }
  messagesCount.forEach(lifetime) { count ->
    val repliesLink = LinkLabel.create(SpaceBundle.message("chat.message.replies.link", count)) {
      SpaceStatsCounterCollector.OPEN_THREAD.log(project)
      action()
    }
    contentPanel.removeAll()
    contentPanel.addToCenter(repliesLink)
    contentPanel.revalidate()
    contentPanel.repaint()
  }
  return contentPanel
}