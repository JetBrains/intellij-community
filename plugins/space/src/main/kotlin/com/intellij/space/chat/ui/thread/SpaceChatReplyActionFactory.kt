// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.space.chat.ui.thread

import circlet.m2.channel.M2ChannelVm
import com.intellij.space.chat.ui.SpaceChatAvatarType
import com.intellij.space.chat.ui.SpaceChatItemComponentFactory
import com.intellij.space.chat.ui.createNewMessageField
import com.intellij.space.messages.SpaceBundle
import com.intellij.ui.components.labels.LinkLabel
import com.intellij.util.ui.JBUI
import com.intellij.util.ui.codereview.SingleValueModelImpl
import com.intellij.util.ui.codereview.ToggleableContainer
import javax.swing.JComponent

internal class SpaceChatReplyActionFactory(
  private val avatarType: SpaceChatAvatarType = SpaceChatAvatarType.THREAD
) : SpaceChatThreadActionsFactory {
  override fun createActionsComponent(chatVm: M2ChannelVm): JComponent {
    val newMessageStateModel = SingleValueModelImpl(false)
    val replyActionSupplier: () -> JComponent = {
      val label = LinkLabel<Any>(SpaceBundle.message("chat.reply.action"), null) { _, _ ->
        newMessageStateModel.value = true
      }
      JBUI.Panels.simplePanel(label).apply {
        isOpaque = false
        border = JBUI.Borders.emptyLeft(avatarType.size.get() + SpaceChatItemComponentFactory.Item.AVATAR_GAP)
      }
    }
    return ToggleableContainer.create(
      newMessageStateModel,
      mainComponentSupplier = replyActionSupplier,
      toggleableComponentSupplier = {
        createNewMessageField(chatVm, onCancel = { newMessageStateModel.value = false }, avatarType = avatarType)
      }
    )
  }
}