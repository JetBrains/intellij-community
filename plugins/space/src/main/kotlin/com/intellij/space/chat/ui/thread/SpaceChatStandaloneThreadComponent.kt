// Copyright 2000-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.space.chat.ui.thread

import circlet.client.api.M2ChannelRecord
import circlet.m2.ChannelsVm
import circlet.m2.M2MessageVm
import circlet.m2.channel.M2ChannelVm
import circlet.platform.api.Ref
import com.intellij.openapi.Disposable
import com.intellij.openapi.project.Project
import com.intellij.space.chat.model.api.SpaceChatItem
import com.intellij.space.chat.model.impl.SpaceChatItemImpl.Companion.convertToChatItem
import com.intellij.space.chat.ui.SpaceChatContentPanelBase
import com.intellij.space.chat.ui.getLink
import com.intellij.space.stats.SpaceStatsCounterCollector
import libraries.coroutines.extra.Lifetime

internal class SpaceChatStandaloneThreadComponent(
  private val project: Project,
  private val lifetime: Lifetime,
  parent: Disposable,
  channelsVm: ChannelsVm,
  chatRecord: Ref<M2ChannelRecord>,
  private val pendingStateProvider: () -> Boolean,
  private val threadActionsFactory: SpaceChatThreadActionsFactory,
  private val statsPlace: SpaceStatsCounterCollector.SendMessagePlace = SpaceStatsCounterCollector.SendMessagePlace.THREAD,
  private val messageConverter: (index: Int, message: M2MessageVm) -> SpaceChatItem = { _, message ->
    message.convertToChatItem(message.getLink())
  }
) : SpaceChatContentPanelBase(lifetime, parent, channelsVm, chatRecord) {
  override fun onChatLoad(chatVm: M2ChannelVm) {
    stopLoadingContent(
      createThreadComponent(project, lifetime, chatVm, pendingStateProvider, threadActionsFactory, statsPlace = statsPlace,
                            messageConverter = messageConverter)
    )
  }
}