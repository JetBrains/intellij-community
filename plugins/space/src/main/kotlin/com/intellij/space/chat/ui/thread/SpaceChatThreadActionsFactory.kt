// Copyright 2000-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.space.chat.ui.thread

import javax.swing.JComponent

internal interface SpaceChatThreadActionsFactory {
  fun createActionsComponent(): JComponent
}