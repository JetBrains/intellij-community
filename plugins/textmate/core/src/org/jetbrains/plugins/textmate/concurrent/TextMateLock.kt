package org.jetbrains.plugins.textmate.concurrent

interface TextMateLock {
  fun <T> withLock(body: () -> T): T
}