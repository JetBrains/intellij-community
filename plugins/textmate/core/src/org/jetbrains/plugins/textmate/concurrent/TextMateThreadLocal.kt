package org.jetbrains.plugins.textmate.concurrent

internal interface TextMateThreadLocal<T> {
  fun get(): T
  fun set(value: T)
}