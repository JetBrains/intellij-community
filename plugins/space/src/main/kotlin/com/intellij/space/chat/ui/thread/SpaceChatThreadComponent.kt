// Copyright 2000-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.space.chat.ui.thread

import circlet.m2.M2MessageVm
import circlet.m2.channel.M2ChannelVm
import com.intellij.ide.plugins.newui.HorizontalLayout
import com.intellij.ide.plugins.newui.VerticalLayout
import com.intellij.openapi.project.Project
import com.intellij.space.chat.model.api.SpaceChatItem
import com.intellij.space.chat.model.impl.SpaceChatItemImpl.Companion.convertToChatItem
import com.intellij.space.chat.ui.*
import com.intellij.space.messages.SpaceBundle
import com.intellij.space.stats.SpaceStatsCounterCollector
import com.intellij.space.ui.SpaceAvatarProvider
import com.intellij.ui.components.ActionLink
import com.intellij.util.ui.JBUI
import com.intellij.util.ui.codereview.SingleValueModelImpl
import com.intellij.util.ui.codereview.ToggleableContainer
import com.intellij.util.ui.codereview.timeline.TimelineComponent
import com.intellij.util.ui.codereview.timeline.thread.TimelineThreadCommentsPanel
import libraries.coroutines.extra.Lifetime
import javax.swing.JComponent
import javax.swing.JPanel

internal fun createThreadComponent(
  project: Project,
  lifetime: Lifetime,
  thread: M2ChannelVm,
  pendingStateProvider: () -> Boolean,
  threadActionsFactory: SpaceChatThreadActionsFactory? = null,
  withFirst: Boolean = true,
  folded: Boolean = true,
  statsPlace: SpaceStatsCounterCollector.SendMessagePlace = SpaceStatsCounterCollector.SendMessagePlace.THREAD,
  messageConverter: (index: Int, message: M2MessageVm) -> SpaceChatItem = { _, message -> message.convertToChatItem(message.getLink()) }
): JComponent {
  val threadComponent = JPanel(VerticalLayout(0)).apply {
    isOpaque = false
  }
  val server = thread.client.server
  val avatarSize = SpaceChatAvatarType.THREAD.size
  val avatarProvider = SpaceAvatarProvider(lifetime, threadComponent, avatarSize)

  val itemsListModel = SpaceChatItemListModel()

  thread.mvms.forEach(lifetime) { messageList ->
    val messages = messageList.messages
    itemsListModel.messageListUpdated(
      messages
        .drop(if (withFirst) 0 else 1)
        .mapIndexed { index, message -> messageConverter(index, message) }
    )
    threadComponent.isVisible = messages.size > if (withFirst) 0 else 1
  }

  val itemComponentFactory = SpaceChatItemComponentFactory(project, lifetime, server, avatarProvider)
  val threadTimeline = if (folded) {
    TimelineThreadCommentsPanel(
      itemsListModel,
      commentComponentFactory = itemComponentFactory::createComponent,
      offset = 0
    ).apply {
      border = JBUI.Borders.emptyBottom(5)
    }
  }
  else {
    TimelineComponent(itemsListModel, itemComponentFactory, offset = 0).apply {
      border = JBUI.Borders.emptyBottom(5)
    }
  }

  val actionsComponent = createActionsComponent(thread, threadActionsFactory, statsPlace, pendingStateProvider)
  return threadComponent.apply {
    add(threadTimeline, VerticalLayout.FILL_HORIZONTAL)
    add(actionsComponent, VerticalLayout.FILL_HORIZONTAL)
  }
}

private fun createActionsComponent(
  thread: M2ChannelVm,
  threadActionsFactory: SpaceChatThreadActionsFactory?,
  statsPlace: SpaceStatsCounterCollector.SendMessagePlace,
  pendingStateProvider: () -> Boolean
): JComponent {
  val newMessageStateModel = SingleValueModelImpl(false)
  val label = ActionLink(SpaceBundle.message("chat.reply.action")) {
    newMessageStateModel.value = true
  }

  val additionalActions = threadActionsFactory?.createActionsComponent()
  val actionsComponent = JPanel(HorizontalLayout(JBUI.scale(5))).apply {
    isOpaque = false
    border = JBUI.Borders.emptyLeft(SpaceChatAvatarType.THREAD.size.get() + SpaceChatItemComponentFactory.Item.AVATAR_GAP)
    add(label)
    additionalActions?.let { add(it) }
  }
  return ToggleableContainer.create(
    newMessageStateModel,
    mainComponentSupplier = { actionsComponent },
    toggleableComponentSupplier = {
      createNewMessageField(
        thread,
        statsPlace = statsPlace,
        onCancel = { newMessageStateModel.value = false },
        avatarType = SpaceChatAvatarType.THREAD,
        pendingStateProvider = pendingStateProvider
      )
    }
  )
}