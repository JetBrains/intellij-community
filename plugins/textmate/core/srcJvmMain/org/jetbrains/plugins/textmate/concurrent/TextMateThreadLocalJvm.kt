package org.jetbrains.plugins.textmate.concurrent

import kotlin.concurrent.withLock

internal fun <T> createTextMateThreadLocalJvm(): TextMateThreadLocal<T> {
  return object : TextMateThreadLocal<T> {
    private val threadLocal = ThreadLocal<T>()

    override fun get(): T {
      return threadLocal.get()
    }

    override fun set(value: T) {
      threadLocal.set(value)
    }
  }
}

internal fun createTextMateLockJvm(): TextMateLock = object : TextMateLock {
  val lock = java.util.concurrent.locks.ReentrantLock()

  override fun <T> withLock(body: () -> T): T {
    return lock.withLock(body)
  }
}