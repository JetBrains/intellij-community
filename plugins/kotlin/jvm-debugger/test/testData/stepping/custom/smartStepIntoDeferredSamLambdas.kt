package smartStepIntoDefferedSamLambda

import kotlin.concurrent.thread

fun interface DoStuffer {
    fun doStuff()
}

typealias FunT = DoStuffer

fun MutableList<FunT>.foo(f: FunT): MutableList<FunT> {
    add(f)
    return this
}

fun MutableList<FunT>.bar(f: FunT): MutableList<FunT> {
    add(f)
    return this
}

fun putLambdaInACollection(list: MutableList<FunT>, f: FunT) {
    list.add(f)
}

fun MutableList<FunT>.invokeAndClear() {
    forEach { it.doStuff() }
    clear()
}

fun main() {
    // STEP_OVER: 1
    //Breakpoint!
    val list = mutableListOf<FunT>()

    // SMART_STEP_INTO_BY_INDEX: 2
    // RESUME: 1
    list.foo { println(1) }
        .bar { println(2) }
        .foo { println(3) }
    list.invokeAndClear()

    // STEP_OVER: 1
    //Breakpoint!
    stopHere()

    // SMART_STEP_INTO_BY_INDEX: 4
    // RESUME: 1
    list.foo { println(1) }
        .bar { println(2) }
        .foo { println(3) }
    list.invokeAndClear()

    // STEP_OVER: 1
    //Breakpoint!
    stopHere()

    // SMART_STEP_INTO_BY_INDEX: 6
    // RESUME: 1
    list.foo { println(1) }
        .bar { println(2) }
        .foo { println(3) }
    list.invokeAndClear()

    // STEP_OVER: 1
    //Breakpoint!
    stopHere()

    // SMART_STEP_INTO_BY_INDEX: 2
    // RESUME: 1
    putLambdaInACollection(list) { println(4) }
    list.invokeAndClear()

    // STEP_OVER: 1
    //Breakpoint!
    stopHere()

    // SMART_STEP_INTO_BY_INDEX: 2
    // RESUME: 1
    putLambdaInACollection(list) { println(4) }
    val thread = thread(true) { list.invokeAndClear() }
    thread.join()
}

fun stopHere() {

}

// IGNORE_K2