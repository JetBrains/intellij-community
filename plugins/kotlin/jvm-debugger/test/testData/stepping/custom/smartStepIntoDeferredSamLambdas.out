LineBreakpoint created at smartStepIntoDeferredSamLambdas.kt:33
LineBreakpoint created at smartStepIntoDeferredSamLambdas.kt:44
LineBreakpoint created at smartStepIntoDeferredSamLambdas.kt:55
LineBreakpoint created at smartStepIntoDeferredSamLambdas.kt:66
LineBreakpoint created at smartStepIntoDeferredSamLambdas.kt:75
Run Java
Connected to the target VM
smartStepIntoDeferredSamLambdas.kt:33
smartStepIntoDeferredSamLambdas.kt:37
smartStepIntoDeferredSamLambdas.kt:37
smartStepIntoDeferredSamLambdas.kt:44
smartStepIntoDeferredSamLambdas.kt:48
smartStepIntoDeferredSamLambdas.kt:49
smartStepIntoDeferredSamLambdas.kt:55
smartStepIntoDeferredSamLambdas.kt:59
smartStepIntoDeferredSamLambdas.kt:61
smartStepIntoDeferredSamLambdas.kt:66
smartStepIntoDeferredSamLambdas.kt:70
smartStepIntoDeferredSamLambdas.kt:70
smartStepIntoDeferredSamLambdas.kt:75
smartStepIntoDeferredSamLambdas.kt:79
smartStepIntoDeferredSamLambdas.kt:79
Disconnected from the target VM

Process finished with exit code 0
